src/runstats.cpp

```diff
@@ -0,0 +1,76 @@
+#include "runstats.hpp"
+
+#include <iostream>
+#include <iomanip>
+
+namespace {
+// Runstats soll auf diese Weise ein Singleton sein
+Runstats runstats;
+}
+
+Runstats::Runstats() {
+    this->root = new Node("**MAIN**");
+    this->current = this->root;
+}
+
+void Runstats::start(const std::string & name) {
+    // Search in current's child for name
+    Node * previous = NULL;
+    for ( Node * child = runstats.current->first; true ; child = child->next) {
+        if ( child == NULL ) {
+            // notfound
+            child = new Node(name);
```

Comment:
Momentan gibt es in diesem PR keine Speicherverwaltung. Die Nodes die hier mit new angelegt werden, werden nie mit delete gelöscht. Das ist eigentlich schlechter C++ Stil. Ich überlege noch etwas, ob mir da eine Alternative einfällt.

Reply:
Hallo Malte,

weil das einzige Runstats-Objekt selbst so lange lebt wie main(), habe ich auf die Deletes der Node verzichtet. Am Ende von main() wird ja der Heap eh freigegeben.
Trotzdem hast du recht, etwas, das ich mit `new` anlege, sollte auch irgendwann mit `delete` wieder freigegeben werden.

Zum Beispiel so:

```c++
Node::~Node() {
    // std::cerr << "deleting Node at " << this << " - " << name  << '\n';
    if (next) {
        delete next;
    }
    if (first) {
        delete first;
    }
}
```

Mit
```c++
delete runstats.root;
```
würden dann alle verlinkten Nodes weggeräumt.

Die runstats-Routine ist ein Clone einer Routine, die ich vor Jahren in Mainframe-Assembler geschrieben habe, dort werden die Nodes aus Effizienzgründen in einem eigenen vor-allokierten Pool (eine einfache Tabelle mit Platz für ein paar 1000 Nodes) angelegt. Bei Start/Stop werden auch Verweilzeit und verbrauchte CPU-Zeit erfasst und im aktuellen Node akkumuliert. Instrumentiert sind damit dann allgemein verwendete "Hauskomponenten" in den Batch-Cobol-Anwendungen, aber auch SQL- , MQ- und andere interessante Funktionsaufrufe.
In der Firma läuft das ganze auch in der Produktion mit und hat dort schon einige Diskussionen über die Ursache von langen Laufzeiten radikal abgekürzt.

In den Str8ts-Solver habe ich ihn nur aus Neugier eingebaut. So konnte ich leichter erkennen, wie der intern funktioniert.

Soll ich den Destructor noch commit'ten? Oder soll `runstats` und die Option `-t` ganz raus?

Gruß Hans-Peter

Reply:
Grundsätzlich hast du Recht, dass der Speicherplatz nach main wieder freigegeben wird. Wenn man aber irgendwann das Programm erweitert, dass mehrere str8ts gelöst werden sollen, dann kann sich das anhäufen. Zugegebenermaßen auch nicht viel, aber wenn ich ein Ziel für die Zukunft mit diesem Projekt habe, dann ist es im source code einen guten Stil zu pflegen.

Interessant, woher die runstats/Routine kommt.

Eigentlich finde ich runstats ganz gut. Mit delete wäre es natürlich schöner, aber ich bin noch am überlegen wie man es ohne new/delete nur mit der stl lösen könnte. Das fände ich um einiges besser. Wir haben ja auch keinen Zeitdruck... :-)

Gruß, Malte

Reply:
Ich habe inzwischen ein klassische Lösung mit einem Destruktor gebaut (also mit `delete`). 
Ich bin auf [A beginner's look at smart pointers in modern C++](https://www.internalpointers.com/post/beginner-s-look-smart-pointers-modern-c) gestoßen. 
Eine moderne Lösung wären wohl die Verwendung von `std::unique_ptr<Node>` statt `Node *`. Kann ich ja mal ausprobieren, bin ja Rentner.

Reply:
Eine unique_ptr-Lösung fände ich sehr schön!

Reply:
Eine harte Nuss mit den smarten Pointern. 
```c++
struct Node {
    friend class Runstats;
    Node(std::string name, Node * parent) : name(name), parent(parent) { };
    ~Node();
private:
    std::string name;
    std::unique_ptr<Node> first = nullptr;
    std::unique_ptr<Node> next = nullptr;
    Node * parent;
    int count = 1;
    
};
```
`parent` ist ein "observer pointer" geblieben. Die 'new' und 'delete' sind jetzt weg und der Destructor ist nun bis auf die Debug-Nachricht leer:

```c++
Node::~Node() {
    if (opt_debug) {
        std::cerr << "deleting Node at " << this << " - " << name  << '\n';
    }
}
```
Damit der Compile funktioniert, musste ich auf c++17 hochgehen!
Mit c++11 bekomme ich `error: call to implicitly-deleted copy constructor`.

Happy review
Hans-Peter

Reply:
Der copy constructor wurde wohl implizit gelöscht, weil er ill-formed wäre: einen unique pointer kann man nicht mal so einfach kopieren. Man könnte auch shared pointer nehmen oder für den parent einen weak pointer. Auf c++17 hochzugehen ist aber so oder so eine gute Idee.
Ich werde mir den commit morgen genauer ansehen. Danke schon mal fürs Ausprobieren mit den unique pointern!

Reply:
Komisch: sollte eigentlich mit c++17 auch noch implicitly deleted sein.